### PR TITLE
Various internal auth and client changes

### PR DESCRIPTION
**Pull request by chrissnyder**
@rogerhutchings and @edpaget would like different pairs of eyes on this to see if I'm moving in the right direction. Not ready to merge yet.

Now that a number of different custom projects are using the client, here are a number of refactorings that I believe supports custom development a bit better. In no particular order:

- Takes heavy inspiration from https://github.com/zooniverse-ui work on testing and switches to using mocha/chai as a test framework. Not a strict improvement, but I'd personally prefer a consistent test environment for all our JS code.
- `PanoptesClient` now extends `JSONAPIClient`. So no more `client.api` to actually access the client.
- Remove the `checkCurrent()` user promise and falls back to a simple evented system. `client.auth` is an instance of EventEmitter. To catch auth changes, just listen for `change` on `client.auth`
- Isomorphic. Should work fine in both node and the browser.
- Removed all compiled code from the repo. I'm not 100% sure of the ramifications of this.
- Updated auth code to sync back up with PFE.

### CODE REVIEW

File: src/client.js [hunk 1]
```diff
@@ -9,77 +12,33 @@ const DEFAULT_OPTS = {
     'Accept': 'application/vnd.api+json; version=1'
   },
   host: 'https://panoptes.zooniverse.org',
```
Comment (edpaget) [inline src/client.js]: You want to change this to zooniverse.org?

File: gulpfile.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    .pipe(source(config.outputFile))
+    .pipe(gulp.dest(config.outputDir));
+}
+
+gulp.task('build', function(){
+  return bundle();
+});
+
+gulp.task('compile', ['build'] ,function(){
+  process.exit(0);
```
Comment (rogerhutchings) [inline gulpfile.js]: Why not run `gulp build` directly?

**Comment by edpaget** [comment]
How does it work serverside. I think the json-api-client uses XMLHttpRequest which doesn't exist in node as far as I know? We could use something like [superagent](https://github.com/visionmedia/superagent)?

**Comment by rogerhutchings** [comment]
It doesn't, and yeah, @brian-c mentioned using superagent in the client at one point as well. I've been polyfilling it with `xmlhttprequest` and `xmlhttprequest-cookie`

File: gulpfile.js (hunk 2 above)
Comment (edpaget) [inline gulpfile.js]: I think its because `gulp build` won't terminate the gulp process, so it can be reused by the `gulp watch` task.

**Comment by chrissnyder** [comment]
it uses https://github.com/chrissnyder/node-xmlhttprequest-cookie stubbed at https://github.com/zooniverse/panoptes-javascript-client/blob/refactor/src/client.js#L1-L4 to make XMLHttpRequest work in Node. Basically wraps http.request to give it the same API as xmlhttprequest.

Better solution might be to use superagent within json-api-client rather than relying on this polyfill, but I took these changes independent of any changes to json-api-client.

File: src/client.js (hunk 1 above)
Comment (chrissnyder) [inline src/client.js]: Yep, will do

**Comment by edpaget** [comment]
I can't build this. There's not `src/index.js' file which is being used in the gulp tasks.

**Comment by edpaget** [comment]
Ah @chrissnyder you're only installing the XHR polyfill as a devDependency, so it won't be around if you include in a different project.

**Comment by chrissnyder** [comment]
Can now `npm run build` and moved `xmlhttprequest-cookie` to just a straight dependency.

**Comment by chrissnyder** [comment]
@brian-c Been meaning to talk with you about this, but keep getting pulled away on other things. Would still like your thoughts on the direction this is going and maybe it'd be best to put them here so others can chime in as well.

**Comment by brian-c** [comment]
Yeah, my only strong suggestions are to use superagent (like you're already discussing) and start using this in the PFE ASAP.

When v2 of the API is ready we should drop JSON-API-Client for a generic adapter we don't have to maintain. That's going to be a nightmare amount of work, so something to start thinking about.

Nice work!

**Comment by edpaget** [comment]
I just put in a pull request to add support for using superagent to the json-api-client lib: https://github.com/zooniverse/json-api-client/pulls

**Comment by chrissnyder** [comment]
Updated the PR to use Ed's superagent version of JAC. Also did some serious housecleaning of the gulp tasks and dependencies.

**Comment by chrissnyder** [comment]
@edpaget With https://github.com/zooniverse/json-api-client being merged in, I think this PR is also good to go.

**Comment by rogerhutchings** [comment]
Is this ready to go?

**Comment by chrissnyder** [comment]
Will double check this afternoon and report back.

**Comment by brian-c** [comment]
Replaced by #20.

**Pull request closed without merging**